runtime/src/iree/hal/cts/testdata/command_buffer_push_constants_test.mlir

```diff
@@ -0,0 +1,37 @@
+// This program reads a value from an input buffer at the offset specified by
+// a push constant then stores that value into an output buffer.
+
+#pipeline_layout = #hal.pipeline.layout<push_constants = 1, sets = [
+  #hal.descriptor_set.layout<0, bindings = [
+    #hal.descriptor_set.binding<0, storage_buffer>,
+    #hal.descriptor_set.binding<1, storage_buffer>
+  ]>
+]>
+
+hal.executable.source public @executable {
+  hal.executable.export public @extract_value layout(#pipeline_layout) {
+  ^bb0(%arg0: !hal.device):
+    %x, %y, %z = flow.dispatch.workgroup_count_from_dag_root
+    hal.return %x, %y, %z : index, index, index
+  }
+  builtin.module {
+    func.func @extract_value() {
+      // I/O buffers.
+      %c0 = arith.constant 0 : index
+      %in = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) offset(%c0) alignment(32) : !flow.dispatch.tensor<readonly:tensor<4xf32>>
+      %out = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) offset(%c0) alignment(32) : !flow.dispatch.tensor<writeonly:tensor<f32>>
+
+      // Read the input push constant.
+      %push_constant_i32 = hal.interface.constant.load[0] : i32
+
+      // Load from the input buffer at the index in the push constant data.
```

Comment:
you could simplify this by dropping the input buffer and just writing the push constant value to the output in the same type (i32) - this is doing a bit too much and would be more difficult to debug

if you had 4 push constants and then stored them into a tensor<4xi32> you can quickly read that back and compare the values match

Reply:
I'm finding it difficult to write a valid `hal.executable.source` op for use with `--compile-mode=hal-executable`. I was originally aiming for something like what you suggest (read push constants -> write into output buffer) and just tried again without much success...

`hal.interface.constant.load` returns a scalar and `flow.dispatch.tensor.store` only takes tensor inputs. I would need to either convert from a scalar (`i32`) to a tensor (`tensor<i32>` or `tensor<4xi32>`), insert elements one by one into a staging tensor, or store elements one by one into the output tensor. Anything I've tried with `tensor.from_elements`, `flow.tensor.store`, or even `flow.tensor.splat` hits an error here: https://github.com/iree-org/iree/blob/c8ee5717e62a3980a69ff825a7c30da044d720e0/compiler/src/iree/compiler/Codegen/Common/BufferizationAnalysis.cpp#L564-L572

like this:

```mlir
command_buffer_push_constants_test.mlir:31:22: error: 'flow.tensor.splat' op unhandled tensor operation
      %input_splat = flow.tensor.splat %input_0 : tensor<4xi32>
                     ^
command_buffer_push_constants_test.mlir:31:22: note: see current operation: %2 = "flow.tensor.splat"(%1) : (i32) -> tensor<4xi32>
// -----// IR Dump After ConvertToDestinationPassingStyle Failed (iree-codegen-convert-to-destination-passing-style) //----- //
func.func @splat_value() {
  %c0 = arith.constant 0 : index
  %0 = hal.interface.constant.load[0] : i32
  %1 = flow.tensor.splat %0 : tensor<4xi32>
  %2 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) offset(%c0) alignment(32) : !flow.dispatch.tensor<writeonly:tensor<4xi32>>
  flow.dispatch.tensor.store %1, %2, offsets = [0], sizes = [4], strides = [0] : tensor<4xi32> -> !flow.dispatch.tensor<writeonly:tensor<4xi32>>
  return
}
```

Should I be using a dialect other than `flow` or `tensor` for that sort of manipulation, or can I come in at a higher level somehow?

---

I'd like to extend the CTS with more tests exercising individual `hal.` ops like this, so it would be good to figure out a nice workflow for producing focused test cases.

Reply:
I want something like
```mlir
%input_0 = hal.interface.constant.load[0] : i32
%input_1 = hal.interface.constant.load[1] : i32
%input_2 = hal.interface.constant.load[2] : i32
%input_3 = hal.interface.constant.load[3] : i32
%push_constants = tensor.from_elements %input_0, %input_1, %input_2, %input_3 : tensor<4xi32>
flow.dispatch.tensor.store %push_constants, %out, offsets = [0], sizes = [4], strides = [1] : tensor<4xi32> -> !flow.dispatch.tensor<writeonly:tensor<4xi32>>
```

`linalg` ops work but are awkward for writing this sort of test... this splats a single value into a tensor:

```mlir
#pipeline_layout = #hal.pipeline.layout<push_constants = 1, sets = [<0, bindings = [<0, storage_buffer>]>]>
module {
  hal.executable.source public @executable{
    hal.executable.export public @splat_push_constant layout(#pipeline_layout) {
    ^bb0(%arg0: !hal.device, %arg1: index, %arg2: index):
      %x, %y, %z = flow.dispatch.workgroup_count_from_dag_root %arg1, %arg2
      hal.return %x, %y, %z : index, index, index
    }
    builtin.module {
      func.func @splat_push_constant() {
        %c0 = arith.constant 0 : index
        %0 = hal.interface.constant.load[0] : i32
        %1 = tensor.empty() : tensor<4xi32>
        %2 = linalg.fill ins(%0 : i32) outs(%1 : tensor<4xi32>) -> tensor<4xi32>
        %3 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) offset(%c0) alignment(32) : !flow.dispatch.tensor<writeonly:tensor<4xi32>>
        flow.dispatch.tensor.store %1, %3, offsets = [0], sizes = [4], strides = [1] : tensor<4xi32> -> !flow.dispatch.tensor<writeonly:tensor<4xi32>>
        return
      }
    }
  }
```

---

`--compile-mode=hal-executable` uses just `buildHALTransformPassPipeline`: https://github.com/iree-org/iree/blob/07c3c7224b76f800945ebbe5558b0af7ab0fce0e/compiler/src/iree/compiler/API2/Internal/Embed.cpp#L414-L431

brainstorming (each of these feels wrong to me):
* we could run additional input passes prior to that (flow/stream)
* we could support more ops in Codegen/Common/BufferizationAnalysis.cpp
* I could author test cases using a mix of the full pipeline and just the hal pipeline somehow (See what linalg is generated for the simpler IR that I start with? The IR might not end up in an executable though?)

Reply:
Ah, I can use `memref.store`, at least for the CPU target. On GPU targets, I see this error: `op unhandled translation of function without compute ops` since now all the executable does is load push constants and write their values into memory, without any linalg ops.

https://github.com/iree-org/iree/blob/1067b57bef5c7883bd19e8ee6a67079fac33c5dc/compiler/src/iree/compiler/Codegen/SPIRV/KernelConfig.cpp#L1429-L1437 https://github.com/iree-org/iree/blob/1067b57bef5c7883bd19e8ee6a67079fac33c5dc/compiler/src/iree/compiler/Codegen/Utils/Utils.cpp#L591-L603

@MaheshRavishankar or @antiagainst , could one of you advise on a path forward here? I'm trying to add tests exercising specific parts of the HAL, but it looks like I need to make my tests more complex than we'd like?

Reply:
I pushed a (probably sketchy) commit that helps with this new test case.

More context: I'm trying to add a test for HAL driver's implementations of push constants. The HAL C API for using push constants is `iree_hal_command_buffer_push_constants()`, which needs to be used with a command buffer. I decided to extend the [command_buffer_dispatch_test](https://github.com/iree-org/iree/blob/main/runtime/src/iree/hal/cts/command_buffer_dispatch_test.h) in the HAL CTS for this. That test uses `--compile-mode=hal-executable` to compile from a MLIR `hal.executable.source` down to a `hal.executable.binary` that the C++ test code can load directly using `iree_hal_executable_cache_prepare_executable`. I'd like to author a `hal.executable.source` that _just_ uses push constants and does little else, which led to me trying ways to use flow/tensor/memref ops and _not_ linalg ops.

What I found is that the SPIRV and CUDA lowerings both assume that all executables must contain at least one "compute op" (mostly linalg ops) and they use the detected compute ops to define a lowering config with workgroup sizes and such. It seems like I can relax that constraint if I explicitly set `attributes {workgroup_size = [1 : index, 1 : index, 1 : index]}`, though that might break other codegen invariants.